### PR TITLE
[TTAHUB-3035] Programmatic Transaction

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change

Adding a tool to aid in writing tests. It rolls back all modifications to any table in the audit log system from the point of the snapshot until the rollback is called. 

## How to test
Run the tests, only functional in non-production. Performs the opposite db operations in reverse order.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-3035 

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

**Comment by thewatermethod** [comment]
What's the difference between this and wrapping a test in a transaction managed by sequelize?

**Comment by thewatermethod** [comment]
Test doesn't release something (probably needs to let go of the sequelize connection)

Screenshot 2024-06-05 at 11 19 50 AM

File: src/lib/programmaticTransaction.test.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+    await Topic.create({ name: 'Topic B' });
+    const changes = await transactionModule.fetchAndAggregateChanges(snapshot);
+    expect(changes[0].new_row_data.name).toBe('Topic B');
+    expect(changes[1].new_row_data.name).toBe('Topic A');
+    await transactionModule.rollbackToSnapshot(snapshot);
+  });
+  it('should throw error for unknown dml_type', async () => {
+    const snapshot = await transactionModule.captureSnapshot();
+    const maxIds = await transactionModule.fetchMaxIds();
```
Comment (thewatermethod) [inline src/lib/programmaticTransaction.test.js:69]: unused vars

File: src/lib/programmaticTransaction.test.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+      .rejects
+      .toThrow('Unknown dml_type(INVALID_TYPE) for table: Topic');
+
+    // Restore the original function
+    spy.mockRestore(); // This restores the original implementation
+  });
+  it('should log and rethrow the error during reversion of changes', async () => {
+    jest.spyOn(auditLogger, 'error'); // Spy on auditLogger.error if not already done
+
+    const snapshot = await transactionModule.captureSnapshot();
```
Comment (thewatermethod) [inline src/lib/programmaticTransaction.test.js:93]: unused var

File: src/lib/programmaticTransaction.test.js [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+      requester: 'requester',
+      targetPopulations: ['pop'],
+      reason: ['reason'],
+      participants: ['participants'],
+      topics: ['topics'],
+      ttaType: ['type'],
+      version: 2,
+    };
+
+    const draftReport = {
```
Comment (thewatermethod) [inline src/lib/programmaticTransaction.test.js:191]: unused var

**Comment by GarrettEHill** [comment]
> What's the difference between this and wrapping a test in a transaction managed by sequelize?

Trying to debug tests that use a transaction can be painful as there is no way to see what was done. Also how to effectively get the tests to  actually get called within functions to use the transaction is not strait forward. This bypasses all the frustrations, and allows changes to be seen in the audit log.

**Review by thewatermethod — APPROVED** [review]
Can we add a comment with a usage example in the README, please?

**Review by AdamAdHocTeam — APPROVED** [review]

**Review by nvms — APPROVED** [review]
Screenshot 2024-06-05 at 2 22 46 PM

🎉

**Pull request merged**